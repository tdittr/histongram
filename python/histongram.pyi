from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
