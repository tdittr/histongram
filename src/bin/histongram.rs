use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
