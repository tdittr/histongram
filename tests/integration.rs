use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::Histogram;

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a = Histogram::from_iter(APACHE.split_whitespace());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.split_whitespace());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::from_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
