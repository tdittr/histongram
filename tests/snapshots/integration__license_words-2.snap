---
source: tests/integration.rs
expression: m_counts
---
[
  ("OR", 7),
  ("to", 7),
  ("THE", 6),
  ("the", 6),
  ("IN", 4),
  ("OF", 4),
  ("of", 4),
  ("and", 3),
  ("ANY", 2),
  ("FOR", 2),
  ("OTHER", 2),
  ("SOFTWARE", 2),
  ("Software", 2),
  ("copies", 2),
  ("in", 2),
  ("is", 2),
  ("notice", 2),
  ("this", 2),
  ("without", 2),
  ("\"AS", 1),
  ("\"Software\"),", 1),
  ("(the", 1),
  ("A", 1),
  ("ACTION", 1),
  ("AN", 1),
  ("AND", 1),
  ("ARISING", 1),
  ("AUTHORS", 1),
  ("BE", 1),
  ("BUT", 1),
  ("CLAIM,", 1),
  ("CONNECTION", 1),
  ("CONTRACT,", 1),
  ("COPYRIGHT", 1),
  ("DAMAGES", 1),
  ("DEALINGS", 1),
  ("EVENT", 1),
  ("EXPRESS", 1),
  ("FITNESS", 1),
  ("FROM,", 1),
  ("HOLDERS", 1),
  ("IMPLIED,", 1),
  ("INCLUDING", 1),
  ("IS", 1),
  ("IS\",", 1),
  ("KIND,", 1),
  ("LIABILITY,", 1),
  ("LIABLE", 1),
  ("LIMITED", 1),
  ("MERCHANTABILITY,", 1),
  ("NO", 1),
  ("NONINFRINGEMENT.", 1),
  ("NOT", 1),
  ("OTHERWISE,", 1),
  ("OUT", 1),
  ("PARTICULAR", 1),
  ("PROVIDED", 1),
  ("PURPOSE", 1),
  ("Permission", 1),
  ("SHALL", 1),
  ("SOFTWARE.", 1),
  ("Software,", 1),
  ("Software.", 1),
  ("TO", 1),
  ("TORT", 1),
  ("The", 1),
  ("USE", 1),
  ("WARRANTIES", 1),
  ("WARRANTY", 1),
  ("WHETHER", 1),
  ("WITH", 1),
  ("WITHOUT", 1),
  ("a", 1),
  ("above", 1),
  ("all", 1),
  ("and/or", 1),
  ("any", 1),
  ("associated", 1),
  ("be", 1),
  ("charge,", 1),
  ("conditions:", 1),
  ("copy", 1),
  ("copy,", 1),
  ("copyright", 1),
  ("deal", 1),
  ("distribute,", 1),
  ("do", 1),
  ("documentation", 1),
  ("files", 1),
  ("following", 1),
  ("free", 1),
  ("furnished", 1),
  ("granted,", 1),
  ("hereby", 1),
  ("included", 1),
  ("including", 1),
  ("limitation", 1),
  ("merge,", 1),
  ("modify,", 1),
  ("obtaining", 1),
  ("or", 1),
  ("permission", 1),
  ("permit", 1),
  ("person", 1),
  ("persons", 1),
  ("portions", 1),
  ("publish,", 1),
  ("restriction,", 1),
  ("rights", 1),
  ("sell", 1),
  ("shall", 1),
  ("so,", 1),
  ("software", 1),
  ("subject", 1),
  ("sublicense,", 1),
  ("substantial", 1),
  ("use,", 1),
  ("whom", 1),
]
